classes/quiz_settings.php

```diff
@@ -147,6 +152,193 @@ protected static function define_properties() {
+        $config = $this->get('config');
+        if (!empty($config)) {
+            $configkey = config_key::generate($config)->get_hash();
+            $this->set('configkey', $configkey);
+        }
+    }
+
+    /**
+     * Create or update the config string based on the current quiz settings.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     * @throws \coding_exception
+     */
+    private function compute_config() {
+        $settings = $this->to_record();
+        // Get existing config.
+        $config = $this->get('config');
+        // Parse basic settings into a property list.
+        $plist = new property_list($config);
+
+        // Process all settings that are boolean.
+        $this->process_bool_settings($plist, $settings);
+
+        // Process quit settings.
+        $this->process_quit_settings($plist, $settings);
+
+        // Add all the URL filters.
+        $this->process_url_filters($plist, $settings);
+
+        // Export and save the config, ready for DB.
+        $this->set('config', $plist->to_xml());
+    }
+
+    /**
+     * Use the boolean map to add Moodle boolean setting to config PList.
+     *
+     * @param property_list $plist The config PList.
+     * @param \stdClass $settings The quiz settings.
+     */
+    private function process_bool_settings(property_list $plist, \stdClass $settings) {
+        $map = $this->get_bool_seb_setting_map();
+        foreach ($settings as $setting => $value) {
+            if (isset($map[$setting])) {
+                $enabled = $value === 1 ? true : false;
+                $plist->add_element_to_root($map[$setting], new CFBoolean($enabled));
+            }
+        }
+    }
+
+    /**
+     * Turn hashed quit password and quit link into PList strings and add to config PList.
+     *
+     * @param property_list $plist The config PList.
+     * @param \stdClass $settings The quiz settings.
+     */
+    private function process_quit_settings(property_list $plist, \stdClass $settings) {
+        if (!empty($settings->quitpassword) && is_string($settings->quitpassword)) {
+            // Hash quit password.
+            $hashedpassword = hash('SHA256', $settings->quitpassword);
+            $plist->add_element_to_root('hashedQuitPassword', new CFString($hashedpassword));
+        }
+
+        if (!empty($settings->linkquitseb) && is_string($settings->linkquitseb)) {
+            $plist->add_element_to_root('quitURL', new CFString($settings->linkquitseb));
+        }
+    }
+
+    /**
+     * Turn return separated strings for URL filters into a PList array and add to config PList.
+     *
+     * @param property_list $plist The config PList.
+     * @param \stdClass $settings The quiz settings.
+     */
+    private function process_url_filters(property_list $plist, \stdClass $settings) {
+        // Create rules to each expression provided and add to config.
+        $urlfilterrules = [];
```

Comment:
maybe I'm blind, but do we have all this filter rules in the tests?

Reply:
Yeah, it's part of the quiz settings that gets turned into XML.

Reply:
Yeah. I'd test all of them to be sure they work as expected

Reply:
Added tests for the different types of URL filters